replace ethereum provider provision with store

weilbith opened the pull request:
Closes: #497 

Same strategy as for the configuration. We replace all former injection/provision data with proper stores. Additionally we finally started to use some additional advantages of stores. Namely for this case getters which allow us to write some more convenient and better readable code. Though we are just scratching the surface here.

There is a special case here because we need to make the provider instance explicitly a shallow reference in the store. This is necessary because deep inside the ethers.js library there is an issue of using a deeply proxied object. It is not very nice to have such code restrictions/adaptions in our core. Though we were not able to find a better solution for now. Furthermore this is
basically the same solution as before. With more input, we might be able to improve here in future.
Cause by this, the store getters which are based on the `provider` state explicitly state their return type to ensure that the shallow reference gets "resolved" and no user of the getter has to access the `.value` first, as this would be an awkward exception to the rest of all store interactions.

It has been decided to make all utility functions which depend on an established connection to the Ethereum provider to return `undefined` if that is not the case. This simply avoids the complexity to provide special treatment in these simple functions. It is the responsibility of the caller to decide what to do in case of `undefined`.

With the tests for the new store, it has become necessary to write a first mock implementation. Because this mock targets the `EthereumProvider` interface there is a shared mock within the `tests` directory. There is no `__mock__` in the `src` directory because first, there are multiple implementations of the `EthereumProvider`. For some parts of the code it doesn't matter which implementation is used and the tests should not make a decision which of them to use. In case that one will become removed in future, many tests can break. This relates to the second point, that these objects get instantiated directly in the
tests and are not indirectly used by imports of the component under test.

nekhaly commented on frontend/src/services/transactions/request-manager.ts:
```diff
@@ -27,7 +28,7 @@ export async function sendRequestTransaction(
 ): Promise<Request> {
   if (!request.fee) {
-    request.fee = await getRequestFee(signer, request);
+    request.fee = await getRequestFee(request);
```
This should be somehow the requestManagerAddress and not the full request object! I doubt that this condition has ever been verified before so we didn't see the error that will happen in this case :D

weilbith replied:
Yes, this is a type error and in another branch with more advanced linting I removed that part properly.
I feel like I can just close this conversation. 😉

nekhaly reviewed (changes requested):
I think the `isConnected` value in `RequestDialog` is a breaking change and should be resolved before we merge. Except for that, awesome work is here. Thank you!

manuelwedler commented on frontend/src/services/transactions/token.ts:
```diff
@@ -3,7 +3,7 @@ import { BigNumber, Contract } from 'ethers';
+import { useEthereumProvider } from '@/stores/ethereum-provider';
 
 export async function ensureTokenAllowance(
   signer: DeepReadonly<JsonRpcSigner>,
```
Wouldn’t it be more consistent to also access the signer from the provider store?

weilbith replied:
Good catch. I made also use of the store here. 👍🏾 

PS: We **really** need more strong type checks (I have a branch locally), because I forgot to adapt the function call too, which got not reported. 😑

manuelwedler commented on frontend/src/services/transactions/token.ts:
```diff
@@ -21,21 +21,25 @@ export async function ensureTokenAllowance(
+/**
+ * @returns number of decimals of token, undefined if not connected
+ */
+export async function getTokenDecimals(tokenAddress: string): Promise<BigNumber | undefined> {
```
What would be wrong about getting the provider from the store in the place where this function is called and still pass it as a parameter?
I liked that we were not returning undefined values from the functions.

Your initial comment sounds like you already had a discussion about this. Just want to understand the decisions made here.

weilbith replied:
See [this comment](https://github.com/beamer-bridge/beamer/pull/581#issuecomment-1106279404) and the last commit.

weilbith commented on frontend/src/stores/ethereum-provider.ts:
```diff
@@ -0,0 +1,19 @@
+    provider: shallowRef<EthereumProvider | undefined>(undefined),
+  }),
+  getters: {
+    isConnected: (state) => state.provider !== undefined,
```
We had a good meeting together to get this sorted out. The result was to go back to the old `signer` and `provider` checks which can be undefined or be defined. The `isConnected` got mixed up and confused. The problem here is that with the future upcoming multiple ethereum provider options this will change a lot. We will re-evaluate this again once we reached this state.
I pushed the related fixup commit.

weilbith commented on frontend/src/components/RequestDialog.vue:
```diff
@@ -27,7 +27,7 @@
       </Card>
 
-      <div v-if="!ethereumProvider.signer.value">
+      <div v-if="!ethereumProvider.isConnected">
```
Checkout [the comment](https://github.com/beamer-bridge/beamer/pull/581#discussion_r855884393) from the conversation below.

nekhaly reviewed (approved):
All Good Now! Thank you!

weilbith commented:
Alright, Manuel and I also had a call and we discussed the big topic of how (store-)composables should interact with each other. This seems to be a very critical decision for how to write high quality code in the frontend with clear rules. This has become even more "interesting" with the current fragile topic of Ethereum connections and signers. Especially in the context of having multiple Ethereum connections to source- and target-roll-up. We went through the whole code base and tried to find the optimal strategy for the team. Here is what we came up with:

- Ethereum provider connections (i.e. JSON-RPC connections) and signers will be separated form each other
  - for the moment this only means that the parameters for the below cases will be always split into to, if some function needs both of them, the `ethereum-provider` store remains as is for now
  - the `EthereumProvider` interface and its implementation needs to be revisited for the split between provider and signer
- Pure utility functions (e.g. `getTokenBalance()`, `ensureTokenAllowence()`) should always receive `provider` and `signer` as parameters that must be `!undefined`
- composables do **not** access directly other (store-)composables (i.e. do not import them, using dependency injection only)
- (current) composables can be categorized into three different groups with the following strategies:
  - composables that expose computed values (e.g. `useTokenBalance`, `useGasFee`)
   - they take the `signer` and/or `provider` values as reactive parameters directly in the `use*` function.
   - they must assume that any of the former parameters is `undefined`
   - they expose at least two computed values, where one is the actual value of interest (e.g. token balance)
   - the primary value can be zero or empty if the `provider` or `signer` is not available
   - the second value signals if the first one is really zero/empty or should not be used because of the current state (because not yet connected with a signer for example)
 - composables that wrap an asynchronous function with additional references/computed values for the state, errors, etc.
   - they take the `signer` and/or `provider` values as static parameters to their exposed functions
   - they can assume that all former parameters are always `!undefined`
   - it is the responsibility of the caller to not call this function if a parameter is `undefined` (e.g. not connected), which is enforced by type errors
   - the latter is often handled by Vue components where the function is triggered by a button that is not enabled if the user is not connected no `provider` is available
 - composables which become useless as soon as we try to resolve their direct dependency to stores (e.g. `useChainCheck`)
   - they become simply removed and their code gets inlined where it is used, or exported into a shared utility function
- the usage of the `configuration` store simply follows the same rules with the following addition
  - any utility function or composable should receive (with possible exceptions) the `ChainConfig` instance itself
  - it is the responsibility of the caller to retrieve this information from the store
  - the latter includes the handling of the case where no chain configuration is known (for what ever reason)
  
@manuelwedler please confirm that this is complete and correct.
@nekhaly please review this and criticize or approve.
Thank you! 🙏🏾

manuelwedler commented:
> * any utility function or composable should receive (with possible exceptions) the `ChainConfig` instance itself

Looks good, I'm happy with it. Just one question: With "instance" you mean a reactive variable here, right?

weilbith commented:
> > * any utility function or composable should receive (with possible exceptions) the `ChainConfig` instance itself
> 
> Looks good, I'm happy with it. Just one question: With "instance" you mean a reactive variable here, right?

Can be yes. But I'm not sure if that is necessary. The chain configurations don't change over time. Though we can still do it. I need to checkout how to get references into nested values of the store.

nekhaly commented:
OK.. Reading is not like being in a live discussion. Upon my understanding of the context, we are talking here about the other approach I was discussing with @weilbith at the initial stage of implementing this, and as I don't have arguments against it, and for sure you may have discussed how this would be better than the current approach, then, there's an initial approval on this. Only the concern is about the first point mentioned here!  
If I have understood it well, you are talking about splitting `provider` from `signer` in the store. I couldn't understand the decision background here, because IMO, `signer` will never occur without a `provider`, so, this means that they are tightly-coupled values (in `instantiation` and not in usage). My concern here is, if you didn't discuss that yet, would that separation lead to an invalid state at any point of the app? So, for example, because of any stupid browser behavior, you may have a `signer` from `WalletConnect` while user Is `MetaMask` connected?  
If we can guarantee that this will never happen, so we can go with splitting with no problems, but this's the reason I want to keep this part atomic (`provider` and `signer` instantiated together and die together)!

nekhaly commented:
Also, another thing, and not sure how this would be covered. So, after a successful transaction, we instantiate another `provider` using the target chain `rpcUrl` in configuration to listen to events. In that case, we have two instances of `provider`. How would we tackle this in our state right now, or maybe we shouldn't care about this second `provider` because it's volatile anyway?

weilbith commented:
@nekhaly thanks for reading through and confirming. The good point is that what you are questioning is related to something we will not do now (in this PR) and we will discuss this later all together.

> Also, another thing, and not sure how this would be covered. So, after a successful transaction, we instantiate another `provider` using the target chain `rpcUrl` in configuration to listen to events. In that case, we have two instances of `provider`. How would we tackle this in our state right now, or maybe we shouldn't care about this second `provider` because it's volatile anyway?

Exactly this is the reason why Manuel and I got to the point (plus how WalletConnect works) that we consider to split provider and signer. We will explain you the details of this thinking in a meeting next week. 🙃

weilbith commented:
Alright, I pushed this huge new commit that makes all the changes according to the above design decision.
I'm sorry that this causes so much review work. 🙈 
But I'm very happy that we have now a fundamental decision that seem to work well and which we can follow from now on. I hope you like the code (except the `RequestDialog` and `RequestFormInputs` components.... 😑 ).
While this PR is quite pressuring for all work that has been done in parallel, please take your time, read careful and don't stress yourself.
Thank you very much!!

nekhaly reviewed (approved):
LGTM.. Another approve is there!

weilbith commented:
Notes (to myself):
  - `watchEffect` could be nice for `useTokenBalance` and `useRequestFee`
  - using `toRef` can be used to create reactive references into the configuration store

manuelwedler commented on frontend/src/components/RequestDialog.vue:
```diff
@@ -155,15 +169,22 @@ const submitRequestTransaction = async (formResult: {
     targetAddress: request.targetAddress,
     amount: formResult.amount,
-    fee: feesEther.value,
+    fee: formattedRequestFeeAmount.value, // TODO: Do we need the formatted amount here?!
```
I think this value is not used anywhere and could be removed entirely from the type.

weilbith replied:
Nothing simpler than that. 😬

weilbith replied:
✔️

manuelwedler reviewed (changes requested):
Great work here. Thank you! I'm very happy about the decisions we made. Just go over my minor comments.

weilbith commented on frontend/src/views/Home.vue:
```diff
@@ -16,42 +16,35 @@
+  if (ethereumProvider.provider) {
+    const { chainId } = storeToRefs(ethereumProvider);
+
+    watch(chainId, chainChangeHandler);
```
Good point. This will be removed anyway with the upcoming WalletConnect integration, but I just fixed it properly. In fact we can always watch for the `chainId`. Nothing special about it. Tested it and works.

weilbith replied:
✔️

weilbith commented on frontend/src/composables/useWaitForRequestFulfilment.ts:
```diff
@@ -0,0 +1,35 @@
+    try {
+      const currentBlockNumber = await provider.getBlockNumber();
+      await listenOnFulfillment(provider, request, fillManagerAddress, currentBlockNumber);
+    } catch (error) {
```
🙈 
Thank you!!

weilbith replied:
While searching for an alternative name of the variable (you know I dislike shorts like `err` or similar), I found that the MDN documentation refers to this variable as `exception`. So I used this everywhere. Though I would be really happy if the IDE would give a hint for this. 😑

weilbith replied:
PS: The upcoming unit tests will also help. 😬

weilbith replied:
✔️

weilbith commented on frontend/src/composables/useRequestTransaction.ts:
```diff
@@ -69,70 +36,27 @@ export function useRequestTransaction(ethereumProvider: ShallowRef<Readonly<Ethe
 
-      await sendRequestTransaction(signer, request, requestState);
+      await sendRequestTransaction(provider, signer, request, state);
     } catch (error) {
```
✔️

weilbith commented on frontend/src/composables/useRequestFee.ts:
```diff
@@ -0,0 +1,36 @@
+
+    try {
+      amount.value = await getRequestFee(provider.value, requestManagerAddress.value);
+    } catch (error) {
```
✔️

weilbith commented on frontend/src/composables/useRequestSigner.ts:
```diff
@@ -1,24 +1,22 @@
+
+    try {
+      await provider.requestSigner();
+    } catch (error) {
```
✔️

weilbith commented on frontend/src/composables/useRequestSigner.ts:
```diff
@@ -1,24 +1,22 @@
-  const requestSigner = async () => {
-    requestSignerError.value = '';
-    await ethereumProvider.value.requestSigner();
-    if (!ethereumProvider.value.signer.value) {
```
Hmm. Interesting. Yes it was intentional, but I did not test it. 🙈 
The point is that there is an error in the console by MetaMask. So I thought this should work. But somehow this is not a "real" exception we can catch. 🤷🏾 
I recovered the old code and that works.

weilbith replied:
✔️

weilbith commented on frontend/src/composables/useTokenBalance.ts:
```diff
@@ -1,75 +1,75 @@
+        tokenAddress.value,
+        signerAddress.value,
+      );
+    } catch (error) {
```
✔️

weilbith commented:
Alright guys. Thank you very much for your reviews! 🙏🏾 
I addressed the feedback and corrected all issues you found. 🤐 
There are a couple of fixup commits in the history of this PR. You can use them to check any changes I did in regards to address the feedback (though some made it into standalone commits). As this PR seems to become finished and there is a conflict with the `main` branch, I'll now rebase on top of it and thereby also resolve all fixups. Will do a force push soon.

weilbith commented:
Alright. The CI is green, manual tests are successful, playing around with all functionalities did not show any new issues.
The merging is blocked due to Manuel his change requests. I addressed all requests and I'm 100% positive that they are fine as I just implemented them.  Manuel works only part time and will not be present beginning next week. Since this PR touched many parts of the frontend and other PRs and local branches are dependent on it, I'll go ahead an merge it with my administrator rights. Please feel free to talk to me if you think this was a bad behavior and/or bring it up in the next retro perspective meeting.
